### PR TITLE
Ieee802154 layer adaptation

### PR DESCRIPTION
**Pull request by GwendalLaurent**

### CODE REVIEW

File: src/ieee802154.erl [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+% @param MacLayer: A tuple containing the name of the module to use, the parameters in a map and the initial state of the layer
+% If a mock MAC layer is used, no PHY layer is created
+% @end
+%-------------------------------------------------------------------------------
+-spec create_stack(
+        IeeeParams::tuple(), 
+        MacLayer::{MacModule::module(), MacParams::map(), MacState::tuple()}
+       ) -> ok.
+create_stack(IeeeParams, MacLayer) ->
+    {MacModule, MacState, MacParams} = MacLayer,
```
Comment (sylane) [inline src/ieee802154.erl]: Tuple elements order mismatch with the spec...

Comment (sylane) [reply src/ieee802154.erl]: What is the MacState parameter ?

File: src/ieee802154.erl [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    case MacModule of
+        mac_layer -> grisp:add_device(spi2, pmod_uwb);
```
Comment (sylane) [inline src/ieee802154.erl]: It would be better if you would call some function in the module that would do the proper setup.

**Comment by sylane** [comment]
There should be a behaviour for the layer, and mac_layer should implement it. In addition, that would document the API.

**Comment by sylane** [comment]
The layer should have a function used something like `State = Mod:init(Params)` and all further call should look like `State2 = Mod:fun(State, ...)`

File: src/ieee802154.erl [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+-spec transmition(FrameControl :: #frame_control{}, FrameHeader :: #mac_header{}, Payload :: bitstring()) -> ok.
+transmition(FrameControl, FrameHeader, Payload) -> gen_statem:call(?MODULE, {tx, FrameControl, FrameHeader, Payload}, infinity).
+
+reception() -> 
+    gen_statem:call(?NAME, rx, infinity).
+
+
+%%% gen_statem callbacks
+init(#{mac_layer := MAC}) ->
+    Data = #{cache => #{tx => [], rx => []}, mac_layer => MAC},
```
Comment (sylane) [inline src/ieee802154.erl]: Initialise the layer here.

File: src/ieee802154.erl [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+idle({call, From}, {tx, FrameControl, FrameHeader, Payload}, Data) -> 
+    {next_state, tx, Data, [{next_event, internal, {tx, FrameControl, FrameHeader, Payload, From}}]};
+
+idle({call, From}, rx, Data) -> 
+    {next_state, rx, Data, [{next_event, internal, {rx, From}}]}.
+
+rx(_EventType, {rx, From}, #{mac_layer := MacLayer} = Data) ->
+    io:format("Reception ~n"),
+    case MacLayer:reception() of
```
Comment (sylane) [inline src/ieee802154.erl]: Pass the layer state to the callback.

File: src/ieee802154.erl [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        {FrameControl, FrameHeader, Payload} -> {next_state, idle, Data, [{reply, From, {FrameControl, FrameHeader, Payload}}]};
+        Err -> {next_state, idle, Data, [{reply, From, Err}]}
+    end.
+
+tx(_EventType, {tx,FrameControl, FrameHeader, Payload, From}, #{mac_layer := MacLayer} = Data) -> 
+    case MacLayer:send_data(FrameControl, FrameHeader, Payload) of
```
Comment (sylane) [inline src/ieee802154.erl]: Pass the layer state to the callback.

File: src/ieee802154.erl [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        ok -> {next_state, idle, Data, [{reply, From, ok}]};
+        Err -> {next_state, idle, Data, [{reply, From, Err}]}
+    end.
+
+terminate(_Reason, _State, _Data) ->
+    ok.
```
Comment (sylane) [inline src/ieee802154.erl]: You should probably have a terminate callback in the layer in case it has to do some cleanup.

**Comment by sylane** [comment]
Withe the Mac layer having a state that is passed around like this, it is not a requirement for it to be a gen_server anymore. It only need to be one if it does something active without being triggered by a function call.

**Comment by sylane** [comment]
Could there be multiple mac layers ? Do they have to be restarted independently from the ieee802154 module ? I think I would rather have the Mac layer not be a process for now. And if it was it could probably just be linked to the ieeee802154 process ? In any case, network_sup sounds too generic a name for this.

File: src/gen_mac_layer.erl [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    {Module, Module:init(Params)}.
+
+% ---------------------------------------------------------------------
+% @doc Transmission request to the MAC layer of a MAC frame
+% @end
+% ---------------------------------------------------------------------
+-spec tx(State::term(), FrameControl::#frame_control{}, MacHeader::#mac_header{}, Payload::bitstring()) -> {State::term(), ok} | {State::term(), Error::atom()}.
+tx({Mod, Sub}, FrameControl, MacHeader, Payload) ->
+    case Mod:tx(Sub, FrameControl, MacHeader, Payload) of
+        {ok, Sub2} -> {{Mod, Sub2}, ok};
```
Comment (sylane) [inline src/gen_mac_layer.erl]: I usually would do ` {ok, Sub2} -> {ok, {Mod, Sub2}}` but this is a choice. I am not sure why, but I usually pass the state as the first argument an return it reverse: `foo(Sub, A, B) -> {ok, B2, A2, Sub2} | {error, Reason, Sub2}` 
I have no strong arguments for it though. The only thing maybe would be to keep the `{ok, ...} | {error, ...}` pattern to be kind of consistent.

File: src/gen_mac_layer.erl [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    end.
+
+% ---------------------------------------------------------------------
+% @doc Transmission request to the MAC layer of a MAC frame
+% @end
+% ---------------------------------------------------------------------
+-spec rx(State::term()) -> {State::term(), FrameControl::#frame_control{}, MacHeader::#mac_header{}, Payload::bitstring()} | {State::term(), Error::atom()}.
+rx({Mod, Sub}) ->
+    case Mod:rx(Sub) of
+        {ok, Sub2, {FrameControl, MacHeader, Payload}} -> {{Mod, Sub2}, FrameControl, MacHeader, Payload};
```
Comment (sylane) [inline src/gen_mac_layer.erl]: Same here, I would keek the `{ok, ...} | {error, ...}` pattern.

**Review by sylane — APPROVED** [review]
Looks good !
If I would have a final nitpicking comment, it would be for the state return value position in the result tuple to be consistent between ok result and error result, both last or both first, and for the stop callback to be called terminate, but it is nitpicking, it looks good.

**Comment by GwendalLaurent** [comment]
>  the stop callback to be called terminate

For the stop/terminate of the gen_mac_layer I followed the same architecture as gen_server. `stop` is in the module and `terminate` is the callback

**Pull request merged**